pkg/kubelet/util/pod_startup_latency_tracker.go

```diff
@@ -0,0 +1,179 @@
+
+	if state.firstResourceVersionWhenPodWasSeenAsRunning == 0 {
+		// skip, pod didn't start yet
+		return
+	}
+	if state.recorded {
+		// skip, pod's latency already recorded
+		return
+	}
+
+	podResourceVersion, err := strconv.ParseUint(pod.ResourceVersion, 10, 64)
+	if err != nil {
+		klog.ErrorS(err, "Unable to parse", "pod.ResourceVersion", pod.ResourceVersion)
+		return
+	}
+
+	if podResourceVersion >= state.firstResourceVersionWhenPodWasSeenAsRunning {
+		podStartingDuration := when.Sub(pod.CreationTimestamp.Time)
+		imagePullingDuration := state.lastFinishedPulling.Sub(state.firstStartedPulling)
+		podStartSLOduration := (podStartingDuration - imagePullingDuration).Seconds()
+
+		klog.V(2).InfoS("Observe PodStartSLIDuration metric", "pod", klog.KObj(pod), "podStartSLOduration", podStartSLOduration, "watchObservedTime", when, "pod.CreationTimestamp", pod.CreationTimestamp.Time, "firstStartedPulling", state.firstStartedPulling, "lastFinishedPulling", state.lastFinishedPulling)
+
+		metrics.PodStartSLIDuration.WithLabelValues().Observe(podStartSLOduration)
+		state.recorded = true
+	}
+}
+
+func (p *PodStartupLatencyTracker) RecordImageStartedPulling(podUID types.UID) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[podUID]
+	if state == nil {
+		klog.ErrorS(nil, "ObservedPodOnWatch should happened before image pulling", "pod", podUID)
+		return
+	}
+
+	if state.firstStartedPulling.IsZero() {
+		state.firstStartedPulling = p.clock.Now()
+	}
+}
+
+func (p *PodStartupLatencyTracker) RecordImageFinishedPulling(podUID types.UID) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[podUID]
+	if state == nil {
+		klog.ErrorS(nil, "ObservedPodOnWatch should happened before image pulling", "pod", podUID)
+		return
+	}
+
+	state.lastFinishedPulling = p.clock.Now() // Now is always grater than values from the past.
+}
+
+func (p *PodStartupLatencyTracker) RecordStatusUpdated(pod *v1.Pod) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[pod.UID]
+	if state == nil {
+		klog.ErrorS(nil, "ObservedPodOnWatch should happened before status changes", "pod", klog.KObj(pod))
+		return
+	}
+
+	if state.recorded {
+		// skip, pod latency already recorded
+		return
+	}
+
+	if state.firstResourceVersionWhenPodWasSeenAsRunning != 0 {
+		// skip, pod already started
+		return
+	}
+
+	if hasPodStartedSLO(pod) {
+		version, err := strconv.ParseUint(pod.ResourceVersion, 10, 64)
+		if err != nil {
+			klog.ErrorS(err, "Unable to parse", "pod.ResourceVersion", pod.ResourceVersion)
+			return
+		}
+
+		klog.V(2).InfoS("Mark when the pod was running for the first time", "pod", klog.KObj(pod), "rv", version)
+		state.firstResourceVersionWhenPodWasSeenAsRunning = version
+	}
+}
+
+// hasPodStartedSLO should reflect "Pod startup latency SLI" definition
+// ref: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md
+func hasPodStartedSLO(pod *v1.Pod) bool {
+	if pod.Status.Phase == v1.PodSucceeded {
```

Comment:
I think that, when `pod is Failed`, it is not started in an SLO definition ( this could be added to main doc, for clarification https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md ). So I think that function itself is working as expected.

But that brings a different question - if a Pod Fails, K8s would try to create a new one and the first one, would never reach a Succeeded state. In my case, it looks like a memory leak, as the data in `pods:  map[types.UID]*perPodState{},` would not only never be used - but also, would never be deleted, right?

My proposal - to add:
```
if pod.Status.Phase == v1.PodFailed {
  // it is no longer possible for pod to start, so we can delete tracked data
  DeletePodStartupState(pod.UID)
}
```

wdyt?

Reply:
> But that brings a different question - if a Pod Fails, K8s would try to create a new one and the first one, would never reach a Succeeded state. 

That depends if the pod is backed by a deployment/replica set, but if so, yes, the controller would attempt to create a new pod.

Regarding your proposal, yes, we should also handle Failed pods and delete their state as well because once a pod is Failed it is in a terminal phase and will never change.

Reply:
Added: https://github.com/kubernetes/kubernetes/pull/111930/files#diff-9365914435bee2af0a1f714a83aeb4b8c5ac8b1a4cf8ebe6e4fa83a5039ea31dR62-R66